x-pack/metricbeat/module/mssql/fetcher.go

```diff
@@ -0,0 +1,150 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package mssql
+
+import (
+	"database/sql"
+
+	"github.com/pkg/errors"
+
+	s "github.com/elastic/beats/libbeat/common/schema"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+
+	// Register driver.
+	_ "github.com/denisenkom/go-mssqldb"
+)
+
+// NewFetcher is called from all metricsets to initialize their fetching routines. It opens a different connection to
+// the database for each fetch operation
+func NewFetcher(uri string, qs []string, schema *s.Schema, log *logp.Logger) (*Fetcher, error) {
+	db, err := sql.Open("sqlserver", uri)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create db instance")
+	}
+
+	// Check the connection before executing all queries to reduce the number
+	// of connection errors that we might encounter.
+	if err = db.Ping(); err != nil {
+		return nil, err
+	}
+
+	f := &Fetcher{
+		queries: qs,
+		schema:  schema,
+		db:      db,
+		log:     log,
+	}
+
+	return f, nil
+}
+
+// Fetcher will make queries concurrently to the database to fetch results. It
+// must be created by each metricset and fed with the queries that must be
+// executed, a SQL implementor and a pointer to the schema to "apply". A
+// metricset could, potentially, need more than one query to fill all its
+// results.
+type Fetcher struct {
+	schema *s.Schema // Schema is the metricset schema to apply to the fetched data to turn it into a result mapstr.
+	db     *sql.DB   // Database to execute queries against.
+	log    *logp.Logger
+
+	queries []string // List of queries to execute concurrently.
+}
+
+// Close closes db connection to the server
+func (f *Fetcher) Close() error {
+	return f.db.Close()
+}
+
+// Report receives a mb.ReporterV2 to send the data to the outputs. Because the operations are common between metricsets
+// it can be extracted to a common function
+func (f *Fetcher) Report(reporter mb.ReporterV2) {
+	var err error
+	for _, q := range f.queries {
+		if err = f.getEventsWithQuery(f.db, q, reporter); err != nil {
+			logp.Error(err)
```

Comment:
Why are you passing the reporter to GetEventsWithQuery and further down instead of just reporting and error here on the top level the same as you do with the logging?

Reply:
I suggested that on the concurrent implementation to report the events as soon as they are generated, instead of accumulating them in a shared list that required mutexes.
I still think it can make sense to report the events as soon as they are generated, but it could be more discussed.

Reply:
For context, the original thread https://github.com/elastic/beats/pull/9202/files#r235994447

Reply:
One per row in the response to the query, for the case of the db metricset I guess this will be one per database, this would be something to be added to the documentation of the metricset.

Reply:
1 per db make sense.